README.md
# [WolcenDB](https://wolcendb.erosson.org)

A website with datamined [Wolcen](https://wolcengame.com) item/skill information, in the spirit of [poedb](https://poedb.tw) and the [PoE wiki](https://pathofexile.gamepedia.com/).
Built with [Elm](https://elm-lang.org), [create-elm-app](https://github.com/halfzebra/create-elm-app), [Node](https://nodejs.org/) and [Yarn](https://yarnpkg.com/).

Special thanks to [gabriel-dehan's WolcenExtractor](https://github.com/gabriel-dehan/WolcenExtractor), which does the datamining! Requires [Ruby](https://rubyinstaller.org/downloads/).


## Build and Run

You must install [Node](https://nodejs.org/) and [Yarn](https://yarnpkg.com/).

Run the website with already-exported Wolcen data in `/datamine` - the bare minimum:

* `yarn` installs our dependencies and does general one-time setup
* `yarn start` runs the website
* `yarn build` builds a production version of the website
* `yarn test` runs the unit tests, verifies that Datamine is valid

More details in [CONTRIBUTING.md](CONTRIBUTING.md)!

See also the autogenerated [README-create-elm-app.md](README-create-elm-app.md)!


CHANGELOG.md
Changelog
=========

---

#### 2023-06-16
- Removed links to the inactive community website.

#### 2023-04-18
- Magic affixes now show drop frequency.
- You can now filter magic affixes by drop frequency. Unavailable affixes are hidden by default.
  - Previously, WolcenDB showed some affixes that were in the game's files, but could never drop. Sorry for the confusion!
- Viewing data from old versions of Wolcen works again.

#### 2023-03-26
- Add ultimate-tier unique weapons (level 80).

#### 2023-03-22
- Updated to Wolcen v1.1.7.0, Endgame. [Wolcen v1.1.0.0 to v1.1.7.0 source differences.](https://github.com/erosson/wolcendb/commit/0aa1a7a5e73eb605b60fda01c6fafb25a0c2b2b7)

#### 2020-12-12
- Added Hunt (Tributary) affixes, along with their possible negative effects, to the affixes and item pages.

#### 2020-12-03
- Updated to Wolcen v1.1.0.0, Bloodtrail. [Wolcen v1.0.13.0 to v1.1.0.0 source differences.](https://gitlab.com/erosson/wolcendb/-/commit/7748f0660955722207576c4a680a0890ca74d9ac)

#### 2020-04-23
- Updated to Wolcen v1.0.13.0. [Wolcen v1.0.12.0 to v1.0.13.0 source differences.](https://gitlab.com/erosson/wolcendb/-/commit/9ad79454d22621acc16f42f1644b98b078af3a06)

#### 2020-04-10
- Updated to Wolcen v1.0.12.0. [Wolcen v1.0.11.0 to v1.0.12.0 source differences.](https://gitlab.com/erosson/wolcendb/-/commit/ab4c67e338c4016a5fa44a09dcb0960b6b20bf5b)

#### 2020-03-30
- Updated to Wolcen v1.0.11.0. [Wolcen v1.0.9.0 to v1.0.11.0 source differences.](https://gitlab.com/erosson/wolcendb/-/commit/1943f2be9b4b29820fce77da5136a3c61d1024ad) (I missed v1.0.10.0, sorry!)

#### 2020-03-12
- Added skill data for apocalypse forms.
- Removed the tutorial attack skill from the skill list.

#### 2020-03-10
- Updated to Wolcen v1.0.9.0. [Wolcen v1.0.8.0 to v1.0.9.0 source differences: no changes.](https://gitlab.com/erosson/wolcendb/-/commit/3b8b0f24b87a7e8b1822d2e8f29799deb7574693)

#### 2020-03-09
- Fixed a bug with language switching.
- Each affix on the [affixes page](/affix) now links to a list of all items that can spawn it.
- Fixed a short-lived bug with normal item listings.
- On the front page, links at the bottom are now only as wide as they appear. Previously they were the width of the entire page, very easy to accidentally click.
- [Language](/language) and [Wolcen version](/build-revision) URLs have changed. They also now remain part of the URL as you move around the site, and refreshing the page will no longer reset them.

#### 2020-03-06
- Updated to Wolcen v1.0.8.0. [Wolcen v1.0.7.0 to v1.0.8.0 source differences.](https://gitlab.com/erosson/wolcendb/-/commit/7774750ece265015b1360ad62bbaf5efeba987d7)
- You can now load WolcenDB with data from old versions of Wolcen.
  - For example, take a look at [v1.0.4.0's overpowered Bleeding Edge](/skill-variant/player_laceration_variant_11?build_revision=1.0.4.3_ER) or [v1.0.6.0's zero-cooldown Winter's Grasp](/skill/player_frostnova?build_revision=1.0.6.0_ER).
- Most pages don't need to wait for `searchIndex` to finish loading any more.
- A limited, non-interactive version of some pages is now shown before `datamine` has finished loading. This will be most visible if you're on mobile or on a slow network.
- Added language selection. Still beta-quality; expect bugs.

#### 2020-03-04
- Images have been restored to loot/unique/skill lists.
- Normal loot now has a "tier" filter, set to tier 16 by default. (You can still select "all tiers" if you like long lists.)
- Wolcen version is now shown on the home page.
- Created a community website
- Slightly redesigned some homepage links

#### 2020-03-03
- Images removed from loot/unique/skill lists. They're still visible when viewing only a single item/unique/skill.
  - Site bandwidth has been too high; oops. This is a quick fix. I'd like to bring these images back once I've had time to explore other solutions. Sorry!

#### 2020-03-02
- Search results for skill-variants now link to the skill-variant's page, instead of the skill's page.
- [`/affix`](/affix) now displays keywords for collapsed rows.
- [`/affix`](/affix) supports filtering by prefix/suffix.

#### 2020-03-01
- Added ailment base-damage data.

#### 2020-02-28
- Updated to Wolcen v1.0.7.0. [Wolcen v1.0.6.0 to v1.0.7.0 source differences.](https://gitlab.com/erosson/wolcendb/-/commit/8dd1940849aa9db35ab69f689b24c630a1a63031)
- Skill variants now try to display interesting numbers.
  - This isn't very pretty, but it's now the easiest way I know of to see numbers for things like [Bleeding Edge: Despotic Perseverence's damage per ailment, for example](/skill/player_laceration).
  - I'm sure this needs more work. Let me know which skill-variants are still missing numbers you'd like to see!
- Added skill-variant images.
- Added pages for single skill-variants. (`/skill-variant/...`)
- Added leveling effects to each skill.

#### 2020-02-27
- Item affixes are now grouped by prefix/suffix, instead of a single list with all possible affixes.
  - [Thanks to Wujido](https://youtu.be/jcYnamTR4f8) for teaching me that prefix/suffix matter! Previously I assumed they were an old code artifact and made no difference.
- Weapons now show rage per hit.
- Added an [Offline] link to unique items.

#### 2020-02-26
- Updated to Wolcen v1.0.6.0. [Wolcen v1.0.4.0 to v1.0.6.0 source differences.](https://gitlab.com/erosson/wolcendb/-/commit/c7a8fd560914493c6ee8538a53791159040e6f3f)
- Added a loading indicator.
  - Fixed the loading indicator in IE/Edge a few hours later; oops.

#### 2020-02-25
- Redesigned the affix list.
  - Affixes are now grouped by class.
  - You can filter affixes by level, gem, or keyword.
  - Keywords, Sarisel, CraftOnly, and prefix/suffix take up less space.
- Added Seeker and Diplomat city mission rewards.

#### 2020-02-24
- Redesigned the normal loot list.
  - You can filter the loot list by keyword. There's many links to filtered lists at the top - for example, "two-handed axes" or "heavy helmets".

cli-src/ssr-cli.js
 // https://korban.net/posts/elm/2019-04-23-how-i-generate-static-html-from-elm-elmstatic/
const fs = require('fs').promises
const jsdom = require('jsdom')
const datamine = require('../public/datamine.json')
const searchIndex = require('../public/searchIndex.json')
const mkdirp = require('mkdirp')
const path = require('path')
const ncp = require('ncp')
const {promisify} = require('util')
const {Elm} = require('../datamine.tmp/SSRPagesCLI.elm.js')
const _ = require('lodash/fp')
const buildRevisions = require('../public/buildRevisions.json')
const langs = require('../datamine/lang.json')

const SSR_BUILD_DIR = './build-ssr'

main()

function main() {
  promisify(ncp.ncp)('./build', SSR_BUILD_DIR)
  .then(() => mkdirp(SSR_BUILD_DIR+'/no-ssr'))
  .then(() => fs.copyFile(SSR_BUILD_DIR+'/index.html', SSR_BUILD_DIR+'/no-ssr/index.html'))
  .then(() => Promise.all([
    fs.readFile(SSR_BUILD_DIR+'/index.html').then(f => f.toString()),
    fs.readFile('./CHANGELOG.md').then(f => f.toString()),
  ]))
  .then(([template, changelog]) => {
    const baseFlags = {buildRevisions, langs, datamine, searchIndex, changelog}
    return Promise.all([
      // fetch a list of pages from SSRPagesCLI
      new Promise((resolve, reject) => {
        const flags = baseFlags
        const pagesApp = Elm.SSRPagesCLI.init({flags})
        pagesApp.ports.ssrCliPages.subscribe(resolve)
      }),
      // initialize a SSRRenderCLI renderer. We initialize this once and reuse it for every render.
      // Rerunning the elm program for every render was simpler, but way too slow!
      // This avoids rebuilding datamine indexes, etc.
      fs.readFile('./datamine.tmp/SSRRenderCLI.elm.js')
      .then(elm => elm.toString())
      .then(elm => {
        const flags = {...baseFlags, url: ''}
        const renderDom = new jsdom.JSDOM(template, {runScripts: 'outside-only'})
        renderDom.window.eval(`${elm}\n\nwindow.app = Elm.SSRRenderCLI.init({node: document.getElementById('root'), flags: ${JSON.stringify(flags)}})`)
        return renderDom
      })
    ])
  })
  .then(([ssrCliPages, renderDom]) => {
    const pages = [
      '', '/changelog', '/privacy', '/affix', '/ailment', '/passive', '/skill', '/city/building_seekers_garrison', '/city/building_trade_assembly', '/gem', '/reagent', '/loot', '/loot/unique',
    // ]
    // startup time is too slow to run the generated pages!
    ].concat(ssrCliPages)
    console.log('page count: ', pages.length)
    // return Promise.all(pages.map(path => run(renderDom, path)))
    //
    // NOPE, too many pages for that. Run in batches instead.
    // https://stackoverflow.com/questions/53964228/how-do-i-perform-a-large-batch-of-promises
    // https://blog.neverendingqs.com/2017/12/02/promises-batches-reduce.html
    const batchSize = 1
    const startTime = Date.now()
    return _.flow(
      _.chunk(batchSize),
      // .convert({cap:false}) passes the map index. https://github.com/lodash/lodash/issues/2387
      _.reduce.convert({cap: false})((chain, batch, i) =>
        chain.then(() => {
          if (i % 50 == 0) console.log('batch', i, 'done: ', i * batchSize, '/', pages.length, ", ", (Date.now() - startTime) / 1000, 'sec')
          return Promise.all(batch.map(path => run(renderDom, path)))
        }),
        Promise.resolve(),
      ),
    )(pages)
  })
  .catch(err => {
    console.error(err && err.slice ? err.slice(-10000) : err)
    process.exit(1)
  })
}
function run(renderDom, path_) {
  const url = "https://wolcendb.erosson.org"+path_
  const dest = SSR_BUILD_DIR + path_ + '/index.html'
  // console.log('dest', dest)
  return staticElm({renderDom, url, dest, path_})
  // return staticElm({renderDom, url, path_}).then(html => console.log(html))
  // return staticElm({renderDom, url, path_}).then(html => console.log(html.length))
}
function staticElm({renderDom, dest, url, path_}) {
  return Promise.resolve().then(() => {
    // send port message to change url; wait for port message saying view is rendered
    const oldHtml = renderDom.window.document.documentElement.outerHTML
    renderDom.window.eval(`app.ports.ssrCliRender.send("${url}")`)
    // renderDom.window.app.ports.ssrCliRender.send(url)
    return new Promise(resolve => {
      // renderDom.window.app.ports.urlChange.subscribe(resolve)
      // NOPE that didn't work. Instead, poll for html changes
      let i = 0
      function pollUntilChanged() {
        if (oldHtml == renderDom.window.document.documentElement.outerHTML) {
          i += 1
          if (i % 100 == 0) console.log('polling...', url, i, oldHtml)
          setTimeout(pollUntilChanged, 1)
        }
        else {
          // console.log('resolved', url, i)
          resolve()
        }
      }
      pollUntilChanged()
    })
  })
  .then(() => {
    // ssr'ed content is non-interactive, since Elm can't access it. Disable interactive elements.
    renderDom.window.eval(`Array.from(document.getElementsByTagName('a')).map(el => {el.setAttribute('href', '#'); el.removeAttribute('target'); el.removeAttribute('onClick');})`)
    renderDom.window.eval(`Array.from(document.getElementsByTagName('input')).map(el => el.setAttribute('readonly', 'readonly'))`)
    renderDom.window.eval(`Array.from(document.getElementsByTagName('textarea')).map(el => el.setAttribute('readonly', 'readonly'))`)
    // return Promise.resolve(renderDom.window.document.getElementById('root').innerHTML)
    return '<!DOCTYPE html>\n' + renderDom.window.document.documentElement.outerHTML
  })
  .then(html => {
    if (dest) {
      return mkdirp(path.dirname(dest))
      .then(() => fs.writeFile(dest, html))
    }
    else {
      return Promise.resolve(html)
    }
  })
}
